services/autoscaling/tests/unit/test_utils_aws.py

```diff
@@ -2,18 +2,219 @@
+from simcore_service_autoscaling.models import Resources
+from simcore_service_autoscaling.utils_aws import (
+    EC2Client,
+    EC2Instance,
+    _compose_user_data,
+    closest_instance_policy,
+    ec2_client,
+    find_best_fitting_ec2_instance,
+    get_ec2_instance_capabilities,
+    start_aws_instance,
+)
+
+
+@pytest.fixture
+def app_settings(
+    app_environment: EnvVarsDict,
+) -> ApplicationSettings:
+    return ApplicationSettings.create_from_envs()
+
+
+def test_ec2_client(app_settings: ApplicationSettings):
+    assert app_settings.AUTOSCALING_EC2_ACCESS
+    with ec2_client(app_settings.AUTOSCALING_EC2_ACCESS) as client:
+        ...
+
+    with pytest.raises(
+        botocore.exceptions.ClientError, match=r".+ AWS was not able to validate .+"
+    ):
+        with ec2_client(app_settings.AUTOSCALING_EC2_ACCESS) as client:
+            client.describe_account_attributes(DryRun=True)
+
+
+def test_ec2_client_with_mock_server(
+    mocked_aws_server_envs: None, app_settings: ApplicationSettings
+):
+    # passes without exception
+    assert app_settings.AUTOSCALING_EC2_ACCESS
+    with ec2_client(app_settings.AUTOSCALING_EC2_ACCESS) as client:
+        client.describe_account_attributes(DryRun=True)
+
+
+def test_get_ec2_instance_capabilities(
+    mocked_aws_server_envs: None,
+    aws_allowed_ec2_instance_type_names: list[str],
+    app_settings: ApplicationSettings,
+):
+    assert app_settings.AUTOSCALING_EC2_ACCESS
+    assert app_settings.AUTOSCALING_EC2_INSTANCES
+    instance_types = get_ec2_instance_capabilities(
+        app_settings.AUTOSCALING_EC2_ACCESS, app_settings.AUTOSCALING_EC2_INSTANCES
+    )
+    assert instance_types
+    assert len(instance_types) == len(
+        app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES
+    )
+
+    # all the instance names are found and valid
+    assert all(
+        i.name in app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES
+        for i in instance_types
+    )
+    for (
+        instance_type_name
+    ) in app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_ALLOWED_TYPES:
+        assert any(i.name == instance_type_name for i in instance_types)
+
+
+def test_find_best_fitting_ec2_instance_with_no_instances_raises():
+    # this shall raise as there are no available instances
+    with pytest.raises(Ec2InstanceNotFoundError):
+        find_best_fitting_ec2_instance(
+            available_ec2_instances=[],
+            resources=Resources(cpus=0, ram=ByteSize(0)),
+        )
+
+
+@pytest.fixture
+def random_fake_available_instances(faker: Faker) -> list[EC2Instance]:
+    list_of_instances = [
+        EC2Instance(
+            name=faker.pystr(),
+            cpus=n,
+            ram=ByteSize(n),
+        )
+        for n in range(1, 30)
+    ]
+    random.shuffle(list_of_instances)
+    return list_of_instances
+
+
+def test_find_best_fitting_ec2_instance_closest_instance_policy_with_resource_0_raises(
```

Comment:
yes that is correct.
How do you want to find that out otherwise than by checking the reservations on the task? that is how docker swarm works and I do not see any other workaround.
The catalog/director ensures there is always a reservation set (even a small default one in case it is missing from the service). Now I do not know if OOIL also checks that? @pcrespov any input on that?